From daf3bc29d2dcf4eb6ae2912e6588c3f5a6cba8d6 Mon Sep 17 00:00:00 2001
From: "Ye.Li" <Ye.Li@freescale.com>
Date: Fri, 10 Jul 2015 19:13:53 +0800
Subject: [PATCH 0234/5242] MLK-11236 mtd: gpmi-nand: Fix nand runtime PM
 issue

commit  e399e54d02e9747f4020921d95a4b44915d74643 from
https://source.codeaurora.org/external/imx/linux-imx.git

Because of the delay of auto suspend, the nand clocks are delayed to
disable when calling the clk_set_rate. This causes the clk_set_rate
failed on some platforms like 6q/6qp, and finally lead the NAND not
working.

Signed-off-by: Ye.Li <Ye.Li@freescale.com>
Signed-off-by: Fugang Duan <B38611@freescale.com>
(cherry picked from commit: 1334dd236d4401d6635accb6c8472d8a5ed088b5)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mtd/nand/raw/gpmi-nand/gpmi-lib.c |    5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/mtd/nand/raw/gpmi-nand/gpmi-lib.c b/drivers/mtd/nand/raw/gpmi-nand/gpmi-lib.c
index 6f00414..28f4a3c 100644
--- a/drivers/mtd/nand/raw/gpmi-nand/gpmi-lib.c
+++ b/drivers/mtd/nand/raw/gpmi-nand/gpmi-lib.c
@@ -474,7 +474,12 @@ void gpmi_nfc_apply_timings(struct gpmi_nand_data *this)
 	void __iomem *gpmi_regs = r->gpmi_regs;
 	unsigned int dll_wait_time_us;
 
+	pm_runtime_get_sync(this->dev);
+	clk_disable_unprepare(r->clock[0]);
 	clk_set_rate(r->clock[0], hw->clk_rate);
+	clk_prepare_enable(r->clock[0]);
+	pm_runtime_mark_last_busy(this->dev);
+	pm_runtime_put_autosuspend(this->dev);
 
 	writel(hw->timing0, gpmi_regs + HW_GPMI_TIMING0);
 	writel(hw->timing1, gpmi_regs + HW_GPMI_TIMING1);
-- 
1.7.9.5

